### DISCUSSION TITLE
"More than one operator + matches these operands" error

### DISCUSSION THREAD

**Discussion by tesla-cat** [nlohmann/json#3953] [General]
- on the error line below i get `"More than one operator = matches these operands" error`, how can i fix this?

```cpp
class TaskRsrc {
   public:
    string r_id;
    float work_hr;
    vector<float> units_rng;
    float units_2;

    TaskRsrc(json data) {
        this->r_id = data["r_id"];
        this->work_hr = data["work_hr"];
        this->units_rng = data["units_rng"];    // i get the error on this line
    }
};
```

**Comment by gregmarr** [comment]
It's better to use explicit get rather than implicit conversions.  

    this->units_rng = data["units_rng"].get<std::vector<float>>();

We recommend you do this to disable them:

    #define JSON_USE_IMPLICIT_CONVERSIONS 0

**Reply by tesla-cat** [reply]
may i ask why?
will this increase the speed of  the compiled program?

**Reply by gregmarr** [reply]
No, but it prevents errors due to ambiguity in the conversion of types.  There have been some really weird errors due to implicit conversions, both to and from the `json` type.

**Comment by tesla-cat** [comment]
- and is below the recommended way to convert nested json into c++ objects? 
- if not, how should i change the code below?

```c++
class TaskRsrc {
   public:
    vector<float> units_rng;

    TaskRsrc(json data) {
        this->units_rng = data["units_rng"].get<vector<float>>();
    }
};

class Task {
   public:
    map<string, TaskRsrc> rsrcs;
 
    Task(json data) {
        for (auto const& [k, v] : data["rsrcs"].get<map<string, json>>()) {
            this->rsrcs.insert({k, TaskRsrc(v)});
        }
    }
};
```

**Reply by gregmarr** [reply]
The recommended way is to use `to_json` and `from_json` functions as described at https://github.com/nlohmann/json#arbitrary-types-conversions.

You can probably even use the macros here:

```
#include <string>
#include <vector>
#include <nlohmann/json.hpp>

class TaskRsrc {
  public:
    std::vector<float> units_rng;
    NLOHMANN_DEFINE_TYPE_INTRUSIVE(TaskRsrc, units_rng);
};

class Task {
  public:
    std::map<std::string, TaskRsrc> rsrcs;
    NLOHMANN_DEFINE_TYPE_INTRUSIVE(Task, rsrcs);
};

int main() {
    Task t;
    nlohmann::json j = t;
    Task t2 = j;
    return 0;
}
```
